storage/bootstrap/bootstrapper/commitlog/source.go

```diff
@@ -121,129 +207,144 @@ func (s *commitLogSource) Read(
+				for start, unmergedBlock := range unmergedBlocks.encoders {
+					startInNano := time.Unix(0, start)
+					block := blocksPool.Get()
+					if len(unmergedBlock) == 0 {
+						shardEmptyErrs[shard]++
+						continue
+					} else if len(unmergedBlock) == 1 {
+						block.Reset(startInNano, unmergedBlock[0].enc.Discard())
+					} else {
+						readers := make([]io.Reader, len(unmergedBlock))
+						for i := range unmergedBlock {
+							stream := unmergedBlock[i].enc.Stream()
+							readers[i] = stream
+							defer stream.Finalize()
+						}
 
-					iter := multiReaderIteratorPool.Get()
-					iter.Reset(readers)
+						iter := multiReaderIteratorPool.Get()
+						iter.Reset(readers)
 
-					var err error
-					enc := encoderPool.Get()
-					enc.Reset(start, blopts.DatabaseBlockAllocSize())
-					for iter.Next() {
-						dp, unit, annotation := iter.Current()
-						encodeErr := enc.Encode(dp, unit, annotation)
-						if encodeErr != nil {
-							err = encodeErr
-							errs++
-							break
+						var err error
+						enc := encoderPool.Get()
+						enc.Reset(startInNano, blopts.DatabaseBlockAllocSize())
+						for iter.Next() {
+							dp, unit, annotation := iter.Current()
+							encodeErr := enc.Encode(dp, unit, annotation)
+							if encodeErr != nil {
+								err = encodeErr
+								shardErrs[shard]++
+								break
+							}
 						}
-					}
-					if iterErr := iter.Err(); iterErr != nil {
-						if err == nil {
-							err = iter.Err()
+						if iterErr := iter.Err(); iterErr != nil {
+							if err == nil {
+								err = iter.Err()
+							}
+							shardErrs[shard]++
 						}
-						errs++
-					}
 
-					iter.Close()
-					for i := range unmergedBlock {
-						unmergedBlock[i].enc.Close()
-					}
+						iter.Close()
+						for i := range unmergedBlock {
+							unmergedBlock[i].enc.Close()
+						}
 
-					if err != nil {
-						continue
-					}
+						if err != nil {
+							continue
+						}
 
-					block.Reset(start, enc.Discard())
+						block.Reset(startInNano, enc.Discard())
+					}
+					blocks.AddBlock(block)
+				}
+				if blocks.Len() > 0 {
+					shardResult.AddSeries(unmergedBlocks.id, blocks)
 				}
-				blocks.AddBlock(block)
 			}
-			if blocks.Len() > 0 {
-				shardResult.AddSeries(unmergedBlocks.id, blocks)
+			if len(shardResult.AllSeries()) > 0 {
+				// Prevent race conditions while updating bootstrapResult from multiple go-routines
+				bootstrapResultLock.Lock()
+				// Shard is a slice index so conversion to uint32 is safe
+				bootstrapResult.Add(uint32(shard), shardResult, nil)
+				bootstrapResultLock.Unlock()
 			}
-		}
-		if len(shardResult.AllSeries()) > 0 {
-			bootstrapResult.Add(shard, shardResult, nil)
-		}
+			<-mergeSemaphore
+			wg.Done()
+		}(shard, unmergedShard)
+	}
+	// Wait for all merge goroutines to complete
+	wg.Wait()
+
+	endMergeTime := time.Now()
+	s.log.Infof("Complete merge took: %fs\n", endMergeTime.Sub(startMergeTime).Seconds())
```

Comment:
emit metrics for each phase rather than logging. The entry point into different bootstrappers (`bootstrap.Process`) already [logs](https://github.com/m3db/m3db/blob/master/storage/bootstrap/process.go#L109) the e2e time per bootstrapper,